third_party/move/move-compiler-v2/src/pipeline/instruction_reordering.rs

```diff
@@ -0,0 +1,771 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! In this module, we implement an intra-block instruction reordering optimization.
+//! The goal of this reordering is to make the code friendlier for generating stack
+//! machine bytecode.
+//!
+//! It is expected that this pass is run as the last transformation in the pipeline.
+//! The prerequisite for this transformation is that all error checks have already been
+//! performed and erroneous code has been rejected.
+//!
+//! In order to perform the reordering of instructions within a block, we construct
+//! an edge-ordered data dependence graph. The data dependencies are use-def edges,
+//! and are ordered based on the order of sources in the use instruction.
+//!
+//! There are a set of constraints that restrict what reorderings are safe to perform.
+//! * True data dependencies (read-after-write) must be respected.
+//! * False data dependencies (write-after-read and write-after-write) must be respected.
```

Comment:
'false data dependencies' confused me a bit. Is that some kind of standard notion? I think it is just a data dependency. Maybe `read-after` and `write-after` as naming?